File: app/src/main/java/com/blueskybone/arkscreen/network/HttpConnectionUtils.java
package com.blueskybone.arkscreen.network;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.Nullable;

import com.hjq.toast.Toaster;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.Map;

import javax.net.ssl.HttpsURLConnection;

public class HttpConnectionUtils {

    private final static String TAG = "HttpConnectionUtils";
//    private static HttpsURLConnection httpsConn;

    public enum RequestMethod {
        GET,
        POST
    }

    static class Response {
        int responseCode;
        String responseContent;

        public Response(int respCode, String content) {
            this.responseCode = respCode;
            this.responseContent = content;
        }
    }

    public static HttpsURLConnection httpResponseConnection(URL url, @Nullable Map<String, String> header, RequestMethod method) {
        try {
            Log.e(TAG, "connection");
            HttpsURLConnection httpsConn = (HttpsURLConnection) url.openConnection();
            httpsConn.setRequestMethod(method.toString());
            httpsConn.setConnectTimeout(10000);
            if (header != null) {
                header.forEach(httpsConn::setRequestProperty);
            }
            httpsConn.setDoOutput(false);
            if (method == RequestMethod.POST) {
                httpsConn.setDoOutput(true);
            }
            httpsConn.setDoInput(true);
            httpsConn.setInstanceFollowRedirects(true);
            httpsConn.connect();
            if (httpsConn.getResponseCode() == HttpURLConnection.HTTP_OK) {
                return httpsConn;
            } else {
                return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public static Response httpResponse(URL url, @Nullable String jsonInput, @Nullable Map<String, String> header, RequestMethod method) throws Exception {
        StringBuilder result = new StringBuilder();
        try {
            HttpsURLConnection httpsConn = (HttpsURLConnection) url.openConnection();
            httpsConn.setConnectTimeout(5000);
            httpsConn.setRequestMethod(method.toString());
            if (header != null) {
                header.forEach(httpsConn::setRequestProperty);
            }
            httpsConn.setDoInput(true);
            httpsConn.setDoOutput(false);
            if (method == RequestMethod.POST) {
                httpsConn.setDoOutput(true);
            }
            if (jsonInput != null) {
                DataOutputStream dataOs = new DataOutputStream(
                        httpsConn.getOutputStream());
                dataOs.writeBytes(jsonInput);
                dataOs.flush();
                dataOs.close();
            }
            httpsConn.connect();
            int respCode = httpsConn.getResponseCode();
            InputStreamReader is;
            if (respCode == HttpURLConnection.HTTP_OK) {
                is = new InputStreamReader(httpsConn.getInputStream());
            } else {
                is = new InputStreamReader(httpsConn.getErrorStream());
            }
            BufferedReader buffer = new BufferedReader(is);
            String inputLine;
            while ((inputLine = buffer.readLine()) != null) {
                result.append(inputLine);
            }
            is.close();
            httpsConn.disconnect();
            return new Response(respCode, result.toString());
        } catch (Exception e) {
            e.printStackTrace();
            throw new Exception(e.getMessage());
        }

    }

    public static void writeToLocal(String destination, InputStream input)
            throws IOException {
        int index;
        byte[] bytes = new byte[1024];
        FileOutputStream downloadFile = new FileOutputStream(destination);
        while ((index = input.read(bytes)) != -1) {
            downloadFile.write(bytes, 0, index);
            downloadFile.flush();
        }
        input.close();
        downloadFile.close();
    }

    private static InputStream getResponseStream(URL url) {
        try {
            HttpsURLConnection httpsConn = (HttpsURLConnection) url.openConnection();
            httpsConn.setConnectTimeout(5000);
            httpsConn.setRequestMethod("GET");
            httpsConn.setDoInput(true);
            httpsConn.setDoOutput(false);
            httpsConn.setInstanceFollowRedirects(true);
            httpsConn.connect();
            int responseCode = httpsConn.getResponseCode();
            if (responseCode == 200) {
                InputStream is = httpsConn.getInputStream();
                httpsConn.disconnect();
                return is;
            } else {
                httpsConn.disconnect();
                return null;
            }
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public static void downloadToLocal(String localPath, URL url) throws IOException {
        InputStream is = getResponseStream(url);
        if (is == null) {
            return;
        }
        writeToLocal(localPath, is);
    }
}

File: app/src/main/java/com/blueskybone/arkscreen/network/HttpConnectionUtils.kt
package com.blueskybone.arkscreen.network

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.Charset
import java.util.concurrent.TimeUnit
import java.util.zip.GZIPInputStream
import javax.net.ssl.HttpsURLConnection

/**
 *   Created by blueskybone
 *   Date: 2025/1/14
 */

class HttpConnectionUtils {
    companion object {
        private const val CONNECT_TIMEOUT = 5000 // 连接超时时间
        private const val READ_TIMEOUT = 5000 // 读取超时时间

        private val okHttpClient:OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS) // 替换你的 CONNECT_TIMEOUT
            .readTimeout(15, TimeUnit.SECONDS)
            .build()

        suspend fun httpResponse(
            url: URL,
            jsonInput: String?,
            header: Map<String, String>,
            method: RequestMethod
        ): Response {
            return try {

                println("=== HTTP 请求信息 ===")
                println("URL: $url")
                println("Method: $method")
                println("Headers: $header")
                println("Request Body: $jsonInput")


                val request = Request.Builder()
                    .url(url)
                    .apply {
                        when (method) {
                            RequestMethod.GET -> get()
                            RequestMethod.POST -> {
                                val body =
                                    jsonInput?.toRequestBody("application/json".toMediaType())
                                post(body ?: "".toRequestBody(null))
                            }
                        }
                        header.forEach { (key, value) ->
                            addHeader(key, value)
                        }
                    }
                    .build()


                val okHttpResponse = withContext(Dispatchers.IO) {
                    okHttpClient.newCall(request).execute()
                }


                println("\n=== HTTP 响应信息 ===")
                println("Response Code: ${okHttpResponse.code}")
                println("Response Message: ${okHttpResponse.message}")
                println("Response Headers:")
                okHttpResponse.headers.forEach { (name, value) ->
                    println("  $name: $value")
                }


                val responseBytes = okHttpResponse.body?.bytes() ?: byteArrayOf()
                println(
                    "\nResponse Raw Bytes (Hex): ${
                        responseBytes.joinToString(" ") {
                            "%02x".format(
                                it
                            )
                        }
                    }"
                )
                println("Response Raw Bytes (Length): ${responseBytes.size} bytes")

                if (okHttpResponse.header("Content-Encoding") == "gzip") {
                    val unzipped =
                        withContext(Dispatchers.IO) {
                            GZIPInputStream(ByteArrayInputStream(responseBytes)).bufferedReader()
                        }.readText()
                    Response(
                        responseCode = okHttpResponse.code,
                        responseContent = unzipped
                    )
                }else{
                    val responseBodyString = try {
                        String(responseBytes, Charset.forName("UTF-8"))
                    } catch (e: Exception) {
                        "无法将响应体解码为UTF-8字符串: ${e.message}"
                    }
                    Response(
                        responseCode = okHttpResponse.code,
                        responseContent = responseBodyString
                    )
                }

            } catch (e: Exception) {
                println("\n=== 请求发生异常 ===")
                println("异常类型: ${e.javaClass.name}")
                println("异常信息: ${e.message}")
                e.printStackTrace()
                Response(responseCode = -1, responseContent = "请求失败: ${e.message}")
            }
        }

        private fun readStream(inputStream: InputStream?): String {
            return inputStream?.bufferedReader()?.use { it.readText() } ?: "No Response"
        }

        suspend fun httpResponseConnection(
            url: URL,
            header: Map<String, String>? = null,
            method: RequestMethod
        ): HttpsURLConnection {
            val httpsConn = withContext(Dispatchers.IO) {
                url.openConnection()
            } as HttpsURLConnection
            httpsConn.apply {
                requestMethod = method.toString()
                connectTimeout = CONNECT_TIMEOUT
                readTimeout = READ_TIMEOUT
                doOutput = method == RequestMethod.POST
                doInput = true
                instanceFollowRedirects = true
            }

            // 设置请求头
            header?.forEach { (key, value) ->
                httpsConn.setRequestProperty(key, value)
            }

            return try {
                withContext(Dispatchers.IO) {
                    httpsConn.connect()
                }
                if (httpsConn.responseCode == HttpURLConnection.HTTP_OK) {
                    httpsConn
                } else {
                    throw HttpException("HTTP error: ${httpsConn.responseCode} - ${httpsConn.responseMessage}")
                }
            } catch (e: Exception) {
                httpsConn.disconnect()
                throw HttpException("Failed to connect: ${e.message}", e)
            }
        }

        private fun getResponseStream(url: URL): InputStream? {
            return try {
                val httpsConn = url.openConnection() as HttpsURLConnection
                httpsConn.connectTimeout = 5000
                httpsConn.requestMethod = "GET"
                httpsConn.doInput = true
                httpsConn.doOutput = false
                httpsConn.instanceFollowRedirects = true
                httpsConn.connect()
                if (httpsConn.responseCode == HttpURLConnection.HTTP_OK) {
                    val isr = httpsConn.inputStream
                    isr
                } else {
                    httpsConn.disconnect()
                    throw Exception("connect error" + httpsConn.responseMessage)
                }
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
                throw Exception(e.message)
            }
        }

        fun downloadToLocal(localPath: String, url: URL) {
            val isr = getResponseStream(
                url
            ) ?: return
            var index: Int
            val bytes = ByteArray(1024)
            val downloadFile = FileOutputStream(localPath)
            while (isr.read(bytes).also { index = it } != -1) {
                downloadFile.write(bytes, 0, index)
                downloadFile.flush()
            }
            isr.close()
            downloadFile.close()
        }

    }

    class HttpException(message: String, cause: Throwable? = null) : Exception(message, cause)
}

//deprecated
//        fun writeToLocal(destination: String?, input: InputStream) {
//            var index: Int
//            val bytes = ByteArray(1024)
//            val downloadFile = FileOutputStream(destination)
//            while (input.read(bytes).also { index = it } != -1) {
//                downloadFile.write(bytes, 0, index)
//                downloadFile.flush()
//            }
//            input.close()
//            downloadFile.close()
//        }

